Userland/Libraries/LibGfx/ImageFormats/JPEGWriter.cpp

```diff
@@ -0,0 +1,502 @@
+                    r -= 16;
+                }
+
+                {
+                    // F.3 - Sequential encoding of a non-zero AC coefficient
+                    auto const ssss = csize(coefficient);
+                    auto const rs = (r << 4) + ssss;
+                    TRY(write_symbol(ac_table.from_input_byte(rs)));
+
+                    if (coefficient < 0)
+                        coefficient -= 1;
+
+                    TRY(m_bit_stream.write_bits<u16>(coefficient, ssss));
+                }
+
+                r = 0;
+            }
+        }
+        return {};
+    }
+
+    static u8 csize(i16 coefficient)
+    {
+        VERIFY(coefficient >= -2047 && coefficient <= 2047);
+        return floor(log2(abs(coefficient))) + 1;
+    };
+
+    QuantizationTable m_luminance_quantization_table {};
+    QuantizationTable m_chrominance_quantization_table {};
+
+    Vector<Macroblock> m_macroblocks {};
+    Array<i16, 3> m_last_dc_values {};
+
+    u64 m_vertical_macroblocks {};
+
+    JPEGBigEndianOutputBitStream m_bit_stream;
+};
+
+ErrorOr<void> add_start_of_image(Stream& stream)
+{
+    TRY(stream.write_value<BigEndian<Marker>>(JPEG_SOI));
+    return {};
+}
+
+ErrorOr<void> add_end_of_image(Stream& stream)
+{
+    TRY(stream.write_value<BigEndian<Marker>>(JPEG_EOI));
+    return {};
+}
+
+ErrorOr<void> add_frame_header(Stream& stream, JPEGEncodingContext const& context, Bitmap const& bitmap)
+{
+    // B.2.2 - Frame header syntax
+    TRY(stream.write_value<BigEndian<Marker>>(JPEG_SOF0));
+
+    // Lf = 8 + 3 × Nf, we only support a single image per frame so Nf = 3
+    TRY(stream.write_value<BigEndian<u16>>(17));
+
+    // P
+    TRY(stream.write_value<u8>(8));
+
+    // Y
+    TRY(stream.write_value<BigEndian<u16>>(bitmap.height()));
+
+    // X
+    TRY(stream.write_value<BigEndian<u16>>(bitmap.width()));
+
+    // Nf, as mentioned earlier, we only support Nf = 3
+    TRY(stream.write_value<u8>(3));
+
+    // Encode 3 components
+    for (u8 i {}; i < 3; ++i) {
+        // Ci
+        TRY(stream.write_value<u8>(i + 1));
+
+        // Hi and Vi
+        TRY(stream.write_value<u8>((1 << 4) | 1));
+
+        // Tqi
+        TRY(stream.write_value<u8>((i == 0 ? context.luminance_quantization_table() : context.chrominance_quantization_table()).id));
+    }
+
+    return {};
+}
+
+ErrorOr<void> add_quantization_table(Stream& stream, QuantizationTable const& table)
+{
+    // B.2.4.1 - Quantization table-specification syntax
+    TRY(stream.write_value<BigEndian<Marker>>(JPEG_DQT));
+
+    // Lq = 2 + 1 * 65
+    TRY(stream.write_value<BigEndian<u16>>(2 + 65));
+
+    // Pq and Tq
+    TRY(stream.write_value<u8>((0 << 4) | table.id));
+
+    for (auto coefficient : table.table)
+        TRY(stream.write_value<u8>(coefficient));
+
+    return {};
+}
+
+ErrorOr<Vector<Vector<u8>, 16>> huffman_table_to_vector(OutputHuffmanTable const& table)
+{
+    Vector<Vector<u8>, 16> output {};
+    TRY(output.try_resize(16));
```

Comment:
Why are you resizing a `Vector` with an inline size of 16... to 16? Shouldn't you use the size of `table.table`?

Reply:
Ah you're right, that way of initializing values always tricks me :^) it's all good then!